crates/egui_kittest/src/lib.rs

```diff
@@ -0,0 +1,167 @@
+    input: egui::RawInput,
+    kittest: kittest::State,
+    output: egui::FullOutput,
+    texture_deltas: Vec<TexturesDelta>,
+    update_fn: Box<dyn FnMut(&egui::Context) + 'a>,
+    event_state: EventState,
+}
+
+impl<'a> Debug for Harness<'a> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        self.kittest.fmt(f)
+    }
+}
+
+impl<'a> Harness<'a> {
+    pub(crate) fn from_builder(
+        builder: &HarnessBuilder,
+        mut app: impl FnMut(&egui::Context) + 'a,
+    ) -> Self {
+        let ctx = egui::Context::default();
+        ctx.enable_accesskit();
+        let mut input = egui::RawInput {
+            screen_rect: Some(builder.screen_rect),
+            ..Default::default()
+        };
+        let viewport = input.viewports.get_mut(&ViewportId::ROOT).unwrap();
+        viewport.native_pixels_per_point = Some(builder.dpi);
+
+        let mut output = ctx.run(input.clone(), &mut app);
+
+        Self {
+            update_fn: Box::new(app),
+            ctx,
+            input,
+            kittest: kittest::State::new(
+                output
+                    .platform_output
+                    .accesskit_update
+                    .take()
+                    .expect("AccessKit was disabled"),
+            ),
+            texture_deltas: vec![mem::take(&mut output.textures_delta)],
+            output,
+            event_state: EventState::default(),
+        }
+    }
+
+    pub fn builder() -> HarnessBuilder {
+        HarnessBuilder::default()
+    }
+
+    /// Create a new Harness with the given app closure.
+    ///
+    /// The ui closure will immediately be called once to create the initial ui.
+    ///
+    /// If you e.g. want to customize the size of the window, you can use [`Harness::builder`].
+    ///
+    /// # Example
+    /// ```rust
+    /// # use egui::CentralPanel;
+    /// # use egui_kittest::Harness;
+    /// let mut harness = Harness::new(|ctx| {
+    ///     CentralPanel::default().show(ctx, |ui| {
+    ///         ui.label("Hello, world!");
+    ///     });
+    /// });
+    /// ```
+    pub fn new(app: impl FnMut(&egui::Context) + 'a) -> Self {
+        Self::builder().build(app)
+    }
+
+    /// Set the size of the window.
+    /// Note: If you only want to set the size once at the beginning,
+    /// prefer using [`HarnessBuilder::with_size`].
+    #[inline]
+    pub fn set_size(&mut self, size: Vec2) -> &mut Self {
+        self.input.screen_rect = Some(Rect::from_min_size(Pos2::ZERO, size));
+        self
+    }
+
+    /// Set the DPI of the window.
+    /// Note: If you only want to set the DPI once at the beginning,
+    /// prefer using [`HarnessBuilder::with_dpi`].
+    #[inline]
+    pub fn set_dpi(&mut self, dpi: f32) -> &mut Self {
+        self.ctx.set_pixels_per_point(dpi);
+        self
+    }
+
+    /// Run a frame.
+    /// This will call the app closure with the current context and update the Harness.
+    pub fn run(&mut self) {
```

Comment:
As mentioned before, I think we should call this `step` and then have a separate `run` that for now calls `step` twice (but in the future would run until "stable")